File: Examples/Image/TransferLearning/TransferLearning.py
# Copyright (c) Microsoft. All rights reserved.

# Licensed under the MIT license. See LICENSE.md file in the project root
# for full license information.
# ==============================================================================

from __future__ import print_function
import numpy as np
import os
from PIL import Image
from cntk.device import set_default_device, gpu
from cntk import load_model, Trainer, UnitType
from cntk.layers import Placeholder, Constant
from cntk.graph import find_by_name, get_node_outputs
from cntk.io import MinibatchSource, ImageDeserializer
from cntk.layers import Dense
from cntk.learner import momentum_sgd, learning_rate_schedule, momentum_schedule
from cntk.ops import input_variable, cross_entropy_with_softmax, classification_error, combine, softmax
from cntk.ops.functions import CloneMethod
from cntk.utils import log_number_of_parameters, ProgressPrinter


################################################
################################################
# general settings
make_mode = False
freeze_weights = False
base_folder = os.path.dirname(os.path.abspath(__file__))
tl_model_file = os.path.join(base_folder, "Output", "TransferLearning.model")
output_file = os.path.join(base_folder, "Output", "predOutput.txt")
features_stream_name = 'features'
label_stream_name = 'labels'
new_output_node_name = "prediction"

# Learning parameters
max_epochs = 20
mb_size = 50
lr_per_mb = [0.2]*10 + [0.1]
momentum_per_mb = 0.9
l2_reg_weight = 0.0005

# define base model location and characteristics
_base_model_file = os.path.join(base_folder, "..", "PretrainedModels", "ResNet_18.model")
_feature_node_name = "features"
_last_hidden_node_name = "z.x"
_image_height = 224
_image_width = 224
_num_channels = 3

# define data location and characteristics
_data_folder = os.path.join(base_folder, "..", "DataSets", "Flowers")
_train_map_file = os.path.join(_data_folder, "6k_img_map.txt")
_test_map_file = os.path.join(_data_folder, "1k_img_map.txt")
_num_classes = 102
################################################
################################################


# Creates a minibatch source for training or testing
def create_mb_source(map_file, image_width, image_height, num_channels, num_classes, randomize=True):
    transforms = [ImageDeserializer.scale(width=image_width, height=image_height, channels=num_channels, interpolations='linear')]
    image_source = ImageDeserializer(map_file)
    image_source.map_features(features_stream_name, transforms)
    image_source.map_labels(label_stream_name, num_classes)
    return MinibatchSource(image_source, randomize=randomize)


# Creates the network model for transfer learning
def create_model(base_model_file, feature_node_name, last_hidden_node_name, num_classes, input_features, freeze=False):
    # Load the pretrained classification net and find nodes
    base_model   = load_model(base_model_file)
    feature_node = find_by_name(base_model, feature_node_name)
    last_node    = find_by_name(base_model, last_hidden_node_name)

    # Clone the desired layers with fixed weights
    cloned_layers = combine([last_node.owner]).clone(
        CloneMethod.freeze if freeze else CloneMethod.clone,
        {feature_node: Placeholder(name='features')})

    # Add new dense layer for class prediction
    feat_norm  = input_features - Constant(114)
    cloned_out = cloned_layers(feat_norm)
    z          = Dense(num_classes, activation=None, name=new_output_node_name) (cloned_out)

    return z


# Trains a transfer learning model
def train_model(base_model_file, feature_node_name, last_hidden_node_name,
                image_width, image_height, num_channels, num_classes, train_map_file,
                num_epochs, max_images=-1, freeze=False):
    epoch_size = sum(1 for line in open(train_map_file))
    if max_images > 0:
        epoch_size = min(epoch_size, max_images)

    # Create the minibatch source and input variables
    minibatch_source = create_mb_source(train_map_file, image_width, image_height, num_channels, num_classes)
    image_input = input_variable((num_channels, image_height, image_width))
    label_input = input_variable(num_classes)

    # Define mapping from reader streams to network inputs
    input_map = {
        image_input: minibatch_source[features_stream_name],
        label_input: minibatch_source[label_stream_name]
    }

    # Instantiate the transfer learning model and loss function
    tl_model = create_model(base_model_file, feature_node_name, last_hidden_node_name, num_classes, image_input, freeze)
    ce = cross_entropy_with_softmax(tl_model, label_input)
    pe = classification_error(tl_model, label_input)

    # Instantiate the trainer object
    lr_schedule = learning_rate_schedule(lr_per_mb, unit=UnitType.minibatch)
    mm_schedule = momentum_schedule(momentum_per_mb)
    learner = momentum_sgd(tl_model.parameters, lr_schedule, mm_schedule, l2_regularization_weight=l2_reg_weight)
    trainer = Trainer(tl_model, (ce, pe), learner)

    # Get minibatches of images and perform model training
    print("Training transfer learning model for {0} epochs (epoch_size = {1}).".format(num_epochs, epoch_size))
    log_number_of_parameters(tl_model)
    progress_printer = ProgressPrinter(tag='Training', num_epochs=num_epochs)
    for epoch in range(num_epochs):       # loop over epochs
        sample_count = 0
        while sample_count < epoch_size:  # loop over minibatches in the epoch
            data = minibatch_source.next_minibatch(min(mb_size, epoch_size-sample_count), input_map=input_map)
            trainer.train_minibatch(data)                                    # update model with it
            sample_count += trainer.previous_minibatch_sample_count          # count samples processed so far
            progress_printer.update_with_trainer(trainer, with_metric=True)  # log progress
            if sample_count % (100 * mb_size) == 0:
                print ("Processed {0} samples".format(sample_count))

        progress_printer.epoch_summary(with_metric=True)

    return tl_model


# Evaluates a single image using the provided model
def eval_single_image(loaded_model, image_path, image_width, image_height):
    # load and format image (resize, RGB -> BGR, CHW -> HWC)
    img = Image.open(image_path)
    if image_path.endswith("png"):
        temp = Image.new("RGB", img.size, (255, 255, 255))
        temp.paste(img, img)
        img = temp
    resized = img.resize((image_width, image_height), Image.ANTIALIAS)
    bgr_image = np.asarray(resized, dtype=np.float32)[..., [2, 1, 0]]
    hwc_format = np.ascontiguousarray(np.rollaxis(bgr_image, 2))

    ## Alternatively: if you want to use opencv-python
    # cv_img = cv2.imread(image_path)
    # resized = cv2.resize(cv_img, (image_width, image_height), interpolation=cv2.INTER_NEAREST)
    # bgr_image = np.asarray(resized, dtype=np.float32)
    # hwc_format = np.ascontiguousarray(np.rollaxis(bgr_image, 2))

    # compute model output
    arguments = {loaded_model.arguments[0]: [hwc_format]}
    output = loaded_model.eval(arguments)

    # return softmax probabilities
    sm = softmax(output[0, 0])
    return sm.eval()


# Evaluates an image set using the provided model
def eval_test_images(loaded_model, output_file, test_map_file, image_width, image_height, max_images=-1, column_offset=0):
    num_images = sum(1 for line in open(test_map_file))
    if max_images > 0:
        num_images = min(num_images, max_images)
    print("Evaluating model output node '{0}' for {1} images.".format(new_output_node_name, num_images))

    pred_count = 0
    correct_count = 0
    np.seterr(over='raise')
    with open(output_file, 'wb') as results_file:
        with open(test_map_file, "r") as input_file:
            for line in input_file:
                tokens = line.rstrip().split('\t')
                img_file = tokens[0 + column_offset]
                probs = eval_single_image(loaded_model, img_file, image_width, image_height)

                pred_count += 1
                true_label = int(tokens[1 + column_offset])
                predicted_label = np.argmax(probs)
                if predicted_label == true_label:
                    correct_count += 1

                np.savetxt(results_file, probs[np.newaxis], fmt="%.3f")
                if pred_count % 100 == 0:
                    print("Processed {0} samples ({1} correct)".format(pred_count, (correct_count / pred_count)))
                if pred_count >= num_images:
                    break

    print ("{0} of {1} prediction were correct {2}.".format(correct_count, pred_count, (correct_count / pred_count)))


if __name__ == '__main__':
    set_default_device(gpu(0))
    # check for model and data existence
    if not (os.path.exists(_base_model_file) and os.path.exists(_train_map_file) and os.path.exists(_test_map_file)):
        print("Please run 'python install_data_and_model.py' first to get the required data and model.")
        exit(0)

    # You can use the following to inspect the base model and determine the desired node names
    # node_outputs = get_node_outputs(load_model(_base_model_file))
    # for out in node_outputs: print("{0} {1}".format(out.name, out.shape))

    # Train only if no model exists yet or if make_mode is set to False
    if os.path.exists(tl_model_file) and make_mode:
        print("Loading existing model from %s" % tl_model_file)
        trained_model = load_model(tl_model_file)
    else:
        trained_model = train_model(_base_model_file, _feature_node_name, _last_hidden_node_name,
                                    _image_width, _image_height, _num_channels, _num_classes, _train_map_file,
                                    max_epochs, freeze=freeze_weights)
        trained_model.save(tl_model_file)
        print("Stored trained model at %s" % tl_model_file)

    # Evaluate the test set
    eval_test_images(trained_model, output_file, _test_map_file, _image_width, _image_height)

    print("Done. Wrote output to %s" % output_file)

File: Tests/EndToEndTests/CNTKv2CSharp/CNTKLibraryCSTrainingTest/TransferLearning.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CNTK.CNTKLibraryCSTrainingTest
{
#pragma warning disable CS0219 // Variable is assigned but its value is never used
    public class TransferLearning
    {
        public static void TrainAndEvaluateTransferLearning(DeviceDescriptor device)
        {
            // string base_model_file = "ResNet_18.model";
            // string feature_node_name = "features";
            // string last_hidden_node_name = "z.x";
            // int[] image_dims = new int[] { 3, 224, 224 };
            string base_model_file = "C:/LiqunWA/cntk/ForkSaved/CNTK_103D.model";
            string feature_node_name = "features";
            string last_hidden_node_name = "second_max";
            int[] image_dims = new int[] { 28, 28, 1};

            //string flower_model_file = "FlowersTransferLearning.model";
            //string flower_results_file = "FlowersPredictions.txt";
            //int flower_num_classes = 102;
            int base_model_num_classes = 10;


            //string animal_model_file = "AnimalsTransferLearning.model";
            //string animal_results_file = "AnimalsPredictions.txt";
            // int animal_num_classes = 0; // ?

            bool freeze = false;

            var image_input = Variable.InputVariable(image_dims, DataType.Float, feature_node_name);
            var label_input = Variable.InputVariable(new int[] { base_model_num_classes }, DataType.Float, "outputs");

            Function base_model = Function.Load(base_model_file, device);
            PrintGraph(base_model.RootFunction, 0);
            var tl_model = create_model(base_model, feature_node_name, last_hidden_node_name, 
                base_model_num_classes, image_input, device, freeze);
            PrintGraph(tl_model.RootFunction, 0);
        }

        static string new_output_node_name = "prediction";

        public static Function create_model(Function base_model, string feature_node_name, 
            string last_hidden_node_name, int num_classes, Variable input_features, 
            DeviceDescriptor device, bool freeze = false)
        {
            Variable old_feature_node = base_model.Arguments.Single(a => a.Name == feature_node_name);
            Function last_node = base_model.FindByName(last_hidden_node_name);
            Variable new_feature_node = CNTKLib.PlaceholderVariable(feature_node_name);

            // Clone the desired layers with fixed weights
            Function cloned_layers = Function.Combine(new List<Variable>() { ((Variable)last_node).Owner }).Clone(
                freeze ? ParameterCloningMethod.Freeze : ParameterCloningMethod.Clone,
                new Dictionary<Variable, Variable>() { { old_feature_node, new_feature_node } });

            Console.WriteLine("cloned_layers");
            PrintGraph(cloned_layers.RootFunction, 0);

            // Add new dense layer for class prediction
            Function feat_norm = CNTKLib.Minus(input_features, Constant.Scalar(DataType.Float, 114.0F));
            Function cloned_out = cloned_layers.ReplacePlaceholders(new Dictionary<Variable, Variable>() { { new_feature_node, feat_norm } });

            Console.WriteLine("cloned_layers after ReplacePlaceholders");
            PrintGraph(cloned_layers.RootFunction, 0);

            Console.WriteLine("cloned_out");
            PrintGraph(cloned_out.RootFunction, 0);
            Function z = TestHelper.Dense(cloned_out, num_classes, device, TestHelper.Activation.None, new_output_node_name);

            Console.WriteLine("z");
            PrintGraph(z.RootFunction, 0);

            return cloned_layers;
        }

        static MinibatchSource create_mb_source(string map_file, int[] image_dims, int num_classes, bool randomize = true)
        {
            DictionaryVector transforms = new DictionaryVector();
            transforms.Add(CNTKLib.ReaderScale(image_dims[2], image_dims[1], image_dims[0], "linear"));

            CNTKDictionary deserializer = CNTKLib.ImageDeserializer(map_file, "label", (uint)num_classes, "image", transforms);
            DictionaryVector deserializers = new DictionaryVector();
            deserializers.Add(deserializer);
            MinibatchSourceConfig minibatchSourceConfig = new MinibatchSourceConfig(deserializers, randomize);
            return CNTKLib.CreateCompositeMinibatchSource(minibatchSourceConfig);
        }

        // Trains a transfer learning model
        static void train_model(string base_model_file, string feature_name, int[] image_dims, int num_classes, string last_hidden_node_name,
            string train_map_file, int max_epochs, int mb_size, double lr_per_mb, bool freeze_weights, double l2_reg_weight,
            double momentum_per_mb, DeviceDescriptor device, int max_images = -1)
        {
            int num_epochs = max_epochs;
            int epoch_size = File.ReadLines(train_map_file).Count(); ;
            if (max_images > 0)
                epoch_size = Math.Min(epoch_size, max_images);
            int minibatch_size = mb_size;

            // Create the minibatch source and input variables
            MinibatchSource minibatch_source = create_mb_source(train_map_file, image_dims, num_classes);
            Variable image_input = Variable.InputVariable(image_dims, DataType.Float, "");
            Variable label_input = Variable.InputVariable(new int[] { num_classes }, DataType.Float, "");

            var featureStreamInfo = minibatch_source.StreamInfo("features");
            var labelStreamInfo = minibatch_source.StreamInfo("labels");

            Function base_model = Function.Load(base_model_file, device);
            // Instantiate the transfer learning model and loss function
            Function tl_model = create_model(base_model, feature_name, last_hidden_node_name,
                num_classes, image_input, device, freeze_weights);
            var ce = CNTKLib.CrossEntropyWithSoftmax(tl_model, label_input);
            var pe = CNTKLib.ClassificationError(tl_model, label_input);

            // Instantiate the trainer object
            var lr_schedule = new TrainingParameterScheduleDouble(lr_per_mb, TrainingParameterScheduleDouble.UnitType.Minibatch);
            var mm_schedule = new MomentumAsTimeConstantScheduleCS(momentum_per_mb);


            bool unitGain = true;
            Learner learner = Learner.MomentumSGDLearner(tl_model.Parameters(), lr_schedule, mm_schedule, unitGain,
                           new AdditionalLearningOptions()
                           {
                               l2RegularizationWeight = l2_reg_weight
                           });
            Trainer trainer = Trainer.CreateTrainer(tl_model, ce, pe, new List<Learner>() { learner });

            // Get minibatches of images and perform model training
            // print("Training transfer learning model for {0} epochs (epoch_size = {1}).".format(num_epochs, epoch_size))
            // CNTKLib.Logging.log_number_of_parameters(tl_model)
            // progress_printer = C.logging.ProgressPrinter(tag = 'Training', num_epochs = num_epochs)
            int outputFrequencyInMinibatches = 1;
            for (int epoch = 0; epoch < num_epochs; epoch++)
            {
                uint sample_count = 0;
                int minibatchIndex = 0;
                while (sample_count < epoch_size)  // loop over minibatches in the epoch
                {
                    uint minibatchSize = (uint)Math.Min(minibatch_size, epoch_size - sample_count);
                    var minibatchData = minibatch_source.GetNextMinibatch(minibatchSize, device);
                    if (minibatchData.empty())
                        break;

                    sample_count += trainer.PreviousMinibatchSampleCount();
                    var arguments = new Dictionary<Variable, MinibatchData>
                    {
                        { image_input, minibatchData[featureStreamInfo] },
                        { label_input, minibatchData[labelStreamInfo] }
                    };
                    trainer.TrainMinibatch(arguments, device);
                    TestHelper.PrintTrainingProgress(trainer, minibatchIndex++, outputFrequencyInMinibatches);
                }
            }
        }
        static void PrintGraph(Function function, int spaces)
        {
            if (function.Inputs == null || function.Inputs.Count() == 0)
            {
                Console.WriteLine(new string('.', spaces) + "(" + function.Name + ")" + function.AsString());
                return;
            }
            
            foreach (var input in function.Inputs)
            {
                Console.WriteLine(new string('.', spaces) + "(" + function.Name + ")" + "->" + 
                    "(" + input.Name + ")" + input.AsString());
            }
            foreach (var input in function.Inputs)
            {
                if (input.Owner != null)
                {
                    Function f = input.Owner;
                    PrintGraph(f, spaces + 4);
                }
            }
        }
    }
}
